hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JobMetricsTest.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.core;
+
+import com.hazelcast.jet.JetInstance;
+import com.hazelcast.jet.Job;
+import com.hazelcast.jet.TestInClusterSupport;
+import com.hazelcast.jet.core.TestProcessors.MockP;
+import com.hazelcast.jet.core.TestProcessors.MockPS;
+import com.hazelcast.jet.core.TestProcessors.NoOutputSourceP;
+import com.hazelcast.jet.core.processor.Processors;
+import com.hazelcast.jet.function.SupplierEx;
+import com.hazelcast.nio.Address;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Function;
+
+import static com.hazelcast.jet.core.Edge.between;
+import static com.hazelcast.jet.core.JobStatus.RUNNING;
+import static com.hazelcast.jet.core.JobStatus.SUSPENDED;
+import static com.hazelcast.jet.core.TestUtil.assertExceptionInCauses;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.peel;
+import static java.util.Collections.singletonList;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(HazelcastSerialClassRunner.class)
+public class JobMetricsTest extends TestInClusterSupport {
```

Comment:
We should add a stress test: one thread will keep restarting the job, another thread will query the metrics. It must never fail. 

Ideally we should also check that the metrics aren't stale. This might be the way:
- use a processor that will increment a counter in a static field
- that field will be exposed using a `@Probe` on the processor class, this works now
- the test will check that it sees monotonically increasing value for this metric. It must never miss this metric once it was present in the result.
Second test might test the same, but instead of restarting it can do `suspend+sleep(100ms)+resume+sleep(100ms)`

Reply:
Agree that these stress tests would be useful. I will prepare them.